spirv-fuzz: Fix to TransformationInlineFunction

afd opened the pull request:
This fixes a problem where TransformationInlineFunction could lead to
distinct instructions having identical unique ids.  It adds a validity
check to detect this problem in general.

Fixes #3911.

Vasniktel commented on source/fuzz/fuzzer.cpp:
```diff
@@ -192,6 +194,24 @@ bool Fuzzer::ApplyPassAndCheckValidity(
     }
+
+    // Check that distinct instructions have different unique ids.
+    std::unordered_map<uint32_t, opt::Instruction*> unique_ids;
```
I think `std::unordered_set` is sufficient here.

afd replied:
You're right, but the map is useful for debugging.  I extended the comment to explain this:

```
    // Check that all instructions have distinct unique ids.  We map each unique
    // id to the first instruction it is observed to be associated with so that
    // if we encounter a duplicate we have access to the previous instruction -
    // this is a useful aid to debugging.
```

Vasniktel commented on test/fuzz/fuzz_test_util.cpp:
```diff
@@ -113,6 +112,23 @@ bool IsValid(spv_target_env env, const opt::IRContext* ir) {
   }
+
+  // Check that all instructions have distinct unique ids.
+  std::unordered_map<uint32_t, opt::Instruction*> unique_ids;
```
Same here.

afd replied:
Same as above, but I unified the code.

Vasniktel commented on tools/fuzz/fuzz.cpp:
```diff
@@ -618,6 +618,34 @@ void DumpShader(spvtools::opt::IRContext* context, const char* filename) {
 
+// Dumps |transformations| to file |filename| in binary format. Useful for
+// interactive debugging.
+void DumpTransformationsBinary(
```
I might be wrong but it seems that these two functions are not used anywhere in the file.

afd replied:
That's right - they're just useful when debugging; in the debugger you can set a breakpoint and then ask for various things to be dumped.

Vasniktel reviewed (commented):
Looks good. It would probably be better to update `fuzzerutil::IsValid` function as well (or maybe refactor it so that this function is used everywhere we need to check the module's consistency).

afd commented:
@Vasniktel following your suggestion I did some reworking of the various parts of the code base where we check validity.

fuzzerutil's "IsValid" has stayed the same - it just invokes the validator.  But I have added a method, IsValidAndWellFormed that uses IsValid but also does the various well-formedness checks on the module that I added recently.

I revamped all the tests to use fuzzerutil::IsValidAndWellFormed, and while I was at it I revamped the tests in the spirit of "include what you use", so that they no longer rely on including "fuzz_test_util.h" to get various other header files.

It's a huge but hopefully straightforward change.

Vasniktel commented on source/fuzz/fuzzer_util.h:
```diff
@@ -151,7 +151,12 @@ uint32_t GetBoundForCompositeIndex(const opt::Instruction& composite_type_inst,
+bool IsValid(const opt::IRContext* context,
+             spv_validator_options validator_options);
+
+bool IsValidAndWellFormed(const opt::IRContext* context,
```
Missing comments.

afd replied:
Fixed.

afd replied:
I also decided to make IsValid take a message consumer, which led to another pretty big change propagating through a lot of files.

Vasniktel commented on test/fuzz/call_graph_test.cpp:
```diff
@@ -274,7 +277,9 @@ TEST(CallGraphTest, FunctionInDegree) {
   const auto context = BuildModule(env, consumer, shader, kFuzzAssembleOption);
-  ASSERT_TRUE(IsValid(env, context.get()));
+  spvtools::ValidatorOptions validator_options;
+  fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                   kConsoleMessageConsumer);
```
Should we ASSERT_TRUE here?

Vasniktel replied:
This issue occurs a lot in tests.

afd replied:
Oh dear - well done for spotting this!  I used regexes to "fix" the tests and obviously messed up.

I believe I have now rectified everything.

Vasniktel reviewed (commented):
Looks good. I've noticed a general issue with missing `ASSERT_TRUE` when `fuzzerutil::IsValidAndWellFormed` is used. I wonder if that's intentional. Also, I've noticed missing doc comments.

Vasniktel commented:
I think it would be good to use a trick from `spirvfuzz_protobufs.h` to get rid of errors in `gtest.h`. The strange thing is, though, that we didn't have those errors before. I'm not sure why.

afd commented:
It would be great if you could review again, @Vasniktel.

And regarding the googletest build problems: the issue is that googletest itself treats warnings as errors, so this is something that will need to be fixed upstream.  For now we are pinning to an older revision of googletest.

Vasniktel commented on source/fuzz/fuzzer_util.h:
```diff
@@ -23,13 +24,45 @@
 // Provides types and global utility methods for use by the fuzzer
 namespace fuzzerutil {
 
+namespace {
```
I'm not sure if we actually need an anonymous namespace here. I'm not against it, though.

paulthomson replied:
Yeah we should probably define these constants once in the cpp file and then declare them extern here without an anonymous namespace. But I see that kConsoleMessageConsumer is used in tests...maybe only in tests? So that could be moved to fuzzer_test_util.

afd replied:
Thanks - I have addressed this in my latest push.
